handler/pkce/handler.go

```diff
@@ -55,7 +58,11 @@ func (c *Handler) HandleAuthorizeEndpointRequest(ctx context.Context, ar fosite.
 	challenge := ar.GetRequestForm().Get("code_challenge")
 	method := ar.GetRequestForm().Get("code_challenge_method")
 
-	if err := c.validate(challenge, method); err != nil {
+	if len(challenge+method) > 0 && !ar.GetClient().IsPublic() && c.DisablePKCEForConfidentialClients {
+		return errors.WithStack(fosite.ErrInvalidRequest.WithHintf(`OAuth 2.0 Client "%s" is registered as a confidential client with a client secret. PKCE is disabled for confidential clients on this server. You can mark a client as public by setting "token_endpoint_auth_method" to "none".`, ar.GetClient().GetID()))
```

Comment:
I'm not sure if it makes sense to forcibly disable PKCE for confidential clients. I was under the impression that all we want to do is not force confidential clients to have PKCE, not the other way around - or did I misunderstand something?

Reply:
Sorry for the late reply.
OK, what you mean is to add a flag, not to "Disable PKCE For Confidential Clients",  but to "Allow No PKCE Validation for Confidential Clients".
With that flag on, 

* if confidential client's authz request includes pkce parameters, do pkce verifications.
* if confidential client's authz request does not include pkce parameters, that's Ok, just skip pkce verification.

Is this understanding correct?

Reply:
Sorry, this completely slipped my radar!

Currently, `EnforcePKCE` forces all clients to use PKCE. However, some installations only want public clients (so e.g. implicit flow) to have PKCE enforced. The flag should probably be either `DisableForcedPKCEForConfidentialClients` or changing the existing flag `EnforcePKCE` to an enum, for example:

- `EnforcePKCE: fosite.EnforcePKCEForAllClients`
- `EnforcePKCE: fosite.EnforcePKCEForPublicClients`

Does that make sense?

Reply:
+1 for enum.